core/src/main/java/io/parsingdata/metal/expression/value/UnaryValueExpression.java

```diff
@@ -54,19 +55,26 @@ public UnaryValueExpression(final ValueExpression operand) {
     }
 
     @Override
-    public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
+    public ImmutableList<Value> eval(final ParseState parseState, final Encoding encoding) {
         return reverse(eval(operand.eval(parseState, encoding), parseState, encoding, new ImmutableList<>()).computeResult());
     }
 
-    private Trampoline<ImmutableList<Optional<Value>>> eval(final ImmutableList<Optional<Value>> values, final ParseState parseState, final Encoding encoding, final ImmutableList<Optional<Value>> result) {
+    private Trampoline<ImmutableList<Value>> eval(final ImmutableList<Value> values, final ParseState parseState, final Encoding encoding, final ImmutableList<Value> result) {
         if (values.isEmpty()) {
             return complete(() -> result);
         }
-        return intermediate(() -> eval(values.tail, parseState, encoding, result.add(values.head.flatMap(value -> eval(value, parseState, encoding)))));
+        return intermediate(() -> eval(values.tail, parseState, encoding, result.add(checkEval(values.head, parseState, encoding))));
     }
 
     public abstract Optional<Value> eval(final Value value, final ParseState parseState, final Encoding encoding);
 
+    private Value checkEval(final Value value, final ParseState parseState, final Encoding encoding) {
```

Comment:
`checkEval` might not be the best name for this method... maybe `checkedEval`? or `safeEval`?

Reply:
Oh, I commented the same above. I was thinking of `safeEval` as well. `doEval`? :stuck_out_tongue_closed_eyes:

Reply:
Well if you _both_ thought of `safeEval(...)` I'm going to assume that's a better choice and go with that! Here and also at `BinaryValueExpression`.